### ISSUE TITLE
pysam fails to install via pip

### ISSUE THREAD

**Issue by pdkaman** [pysam-developers/pysam#1132]
Hello,

I'm surprised this is not on any documentation or setup.  I had some dependencies that were not clear in the setup but once these were installed the pysam install completed.  

Intially pip install pysam failed because it lacked libzip2.  However this package did not install (i.e. sudo apt-get install libzip2).  

Subsequent to this was another error message indicating the package libbz2-dev (maybe this is the same) was a dependency.
After installing this and retrying pip install pysam there was another error due to a dependency on liblzma-dev.

After installing both of this libraries (Ubuntu) the pysam install completed.

I am willing to provide better documentation on these errors if necessary.  However it seems that pysam has some dependencies that are not noted in the setup or not standard in Ubuntu.

**Comment by marcelm** [comment]
Usually, `pip install pysam` should use pre-compiled binary packages (wheels). These are much faster to install, and they don’t require a compiler nor `-dev` packages to be installed. This is how it looks on my Ubuntu 22.04:
```
$ python3.10 -m venv .venv
$ .venv/bin/pip install pysam
Collecting pysam
  Using cached pysam-0.19.1-cp310-cp310-manylinux_2_24_x86_64.whl (15.2 MB)
Installing collected packages: pysam
Successfully installed pysam-0.19.1
$
```
(Excluding the download, this took less than 1 second.)

Without more information, it is hard to say what is going wrong in your case.

**Comment by pettyalex** [comment]
@marcelm I was able to reproduce this on a brand new Ubuntu 20.04 x86_64 machine in a pretty frustrating way.

When I try to install pysam, none of the wheels match and it tries (and fails) to build from source. A verbose install attempt finds: `Skipping link: none of the wheel's tags match: cp38-cp38-manylinux_2_24_x86_64` even though the machine's triple is `x86_64-linux-gnu`, this clearly should be matching.

I think that the version of pip distributed with Ubuntu 20.04 may be too old to match the `manylinux_2_24` tags. I think that it needs `pip 20.3` at a minimum: https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/#manylinux-compatibility-support
![image](https://user-images.githubusercontent.com/3219141/196221693-c00ad6a4-642d-4411-aaaa-568a7b811aef.png)

While Ubuntu 20.04 is shipping pip version:
```
pip --version
pip 20.0.2 from /usr/lib/python3/dist-packages/pip (python 3.8)
```

**Comment by marcelm** [comment]
> I think that the version of pip distributed with Ubuntu 20.04 may be too old to match the `manylinux_2_24` tags.

You’re right, this is very likely the problem. On your side, you can solve this by using a virtual environment with a newer pip version. So something like this:
```
$ python3 -m venv .venv
# source .venv/bin/activate
$ pip install --upgrade pip
$ pip install pysam
```
Virtual environments are great because they allow you to install packages independent of system packages.

**Comment by aryarm** [comment]
I've also encountered this issue and tried nearly every version of pip

when installing pysam `0.20.0` in python3.7, it seemed to come up only with `pip < 20.3`. So we are telling all of our users to install our package with at least `pip >= 20.3` from now on

**Comment by yzliu01** [comment]
@aryarm I have pip 23.0.1 and got problem to install pysam. 
Something like beneath
```
...
× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
WARNING: Ignoring invalid distribution -ip (c:\python310\lib\site-packages)
```
Appreciate any help to get it installed.

**Comment by aryarm** [comment]
@yzliu01, I don't think wheels (pre-compiled binaries) have been published for Windows OS

Are you using [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/install#install-wsl-command)?

**Comment by yzliu01** [comment]
Hi, now I’m confused to tell the difference. I tried both the terminal via
VS Code and the CMD prompt in windows. I got same problem. Probably it’s
not through WSL in my case?

**Comment by yzliu01** [comment]
Hi, many thanks. Exactly, I could manage to install it via VS Code by launching Ubuntu on WSL which I installed before. Glad to know this trick!

**Comment by ahvigil** [comment]
> We use 2_28 for both x86_64 and aarch64; if we made wheels for i686, we would use 2014. These choices were made because they are non-EOL and they are suitable out of the box on Alma & Rocky 9 and Ubuntu 22.04. CentOS 7 and Alma/Rocky 8 ship with such an old pip that there's no choice but to upgrade it first.

I'm running into an interesting special case as a consequence of pysam using a slightly nonstandard cibuildwheel config. My package has a pysam dependency and also uses cibuildwheel for matrix'd wheel builds. The cibuildwheel default image for x86_64 builds is still `manylinux2014`, but as `pysam` only produces `manylinux_2_28` wheels our CI is forced to compile pysam from the source distribution from inside our older `manylinux` container. This requires `yum install`ing a few build requirements and adds a few minutes to each wheel build, or else we would need to bump our x86_64 manylinux container to match whatever pysam uses.

**Comment by jmarshall** [comment]
@ahvigil: If you think this is a nonstandard configuration, please point at clear documentation of conventions for wheel provision. I preferred not to build manylinux2014 as the baseline, because that would pessimise pysam for users of more up-to-date x86_64 platforms that **do** have glibc 2.28+, which should be the increasingly common case.

However we do also provide a manylinux2014 wheel for the Ubuntu 20.04 system Python. If you would like to have another manylinux2014 wheel also provided, you would need to specify the platform(s) (architecture, distribution, which release of that distro, and in particular Python version) that you are interested in.

If this is a matrix of Python versions, it would also be useful if you explained why keeping manylinux2014 as the baseline for your package is a good thing. Alternatively your package could join pysam in the modern world. 😄

**Comment by ahvigil** [comment]
by nonstandard I simply meant non-default. This probably isn't common enough to warrant a special build but just wanted to note this for posterity, in case someone follows the same path as I did figuring out why wheels aren't being used in my CI pipeline and cibuildwheel running on the default configuration.
